### PR TITLE
feat: Implemented #731

### LINKED ISSUE #731: Extension of factoryinsight for time series TARGET vs ACTUAL production
In many production facilities, the currently used and desired diagram is a time series that compares the TARGET with the ACTUAL production.

See figure:
image

As a user of the United Manufacturing Hubs, I would also like to be able to use this visualisation.

To do this, I only want to tell the system the current products produced via an MQTT topic ("/count"?) and in some way also the target speed.
It would also be interesting to be able to set up the chart so that the target speed is extrapolated and, if necessary, the sum of the target production for this shift is displayed in the same chart. 

This feature was requested by a partner company.

### PR DESCRIPTION
**Pull request by Scarjit**
# Description

This pr provides an extension to factoryinsight, for retrieving the actual vs target production.
The latest version of this pr is also on [docker hub](https://hub.docker.com/repository/docker/scarjit/factoryinsight)

Fixes #731 

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [X] Checked HTTP response with Insomnia
- [x] Data appears correctly in Grafana

**Test Configuration**:
* Kubernetes Version: 
* Helm Version:
* Hardware:

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [X] I have checked that my issue complies with the [Contributing guidelines](https://github.com/united-manufacturing-hub/united-manufacturing-hub/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by Scarjit** [comment]
image

->

image

File: golang/cmd/factoryinsight/http.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ import (
 
 // SetupRestAPI initializes the REST API and starts listening
 func SetupRestAPI(accounts gin.Accounts, version string, jaegerHost string, jaegerPort string) {
-	gin.SetMode(gin.ReleaseMode)
+	gin.SetMode(gin.DebugMode)
```
Comment (JeremyTheocharis) [inline golang/cmd/factoryinsight/http.go:21]: Please move it back to Release

File: golang/cmd/factoryinsight/main.go [hunk 2]
```diff
@@ -31,12 +31,24 @@ import (
...
+		if debugEnabled == "True" || debugEnabled == "true" {
+			logger, _ = zap.NewDevelopment()
+		} else {
+			logger, _ = zap.NewProduction()
+		}
 
-	// Setup logger and set as global
-	logger, _ := zap.NewProduction()
+	*/
+	logger, _ := zap.NewDevelopment()
```
Comment (JeremyTheocharis) [inline golang/cmd/factoryinsight/main.go:48]: Is it on purpose development?

File: golang/cmd/factoryinsight/http.go [hunk 3]
```diff
@@ -2044,12 +2046,44 @@ func processUniqueProductsWithTagsRequest(c *gin.Context, getDataRequest getData
...
+	var getProcessAccumulatedProducts getProcessAccumulatedProducts
+	var err error
+
+	err = c.BindQuery(&getProcessAccumulatedProducts)
+	if err != nil {
+		handleInvalidInputError(span, c, err)
+		return
+	}
+
+	accumulatedProducts, err := GetAccumulatedProducts(span, getDataRequest.Customer, getDataRequest.Location, getDataRequest.Asset, getProcessAccumulatedProducts.From, getProcessAccumulatedProducts.To)
```
Comment (JeremyTheocharis) [inline golang/cmd/factoryinsight/http.go:2082]: Where is this function?

File: golang/internal/regression.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package internal
```
Comment (JeremyTheocharis) [inline golang/internal/regression.go:1]: Where is this used?

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]
Did you push all your commits? This looks like it is not fully finished (only went through the code, did not test it)

**Comment by Scarjit** [comment]
> Did you push all your commits? This looks like it is not fully finished (only went through the code, did not test it)

This was finished, until @kruegeralex found some bugs

File: golang/cmd/factoryinsight/database.go [hunk 5]
```diff
@@ -1789,3 +1791,645 @@ func GetUniqueProductsWithTags(parentSpan opentracing.Span, customerID string, l
 
 	return
 }
+
+// GetAccumulatedProducts gets the accumulated counts for an observation timeframe and an asset
```
Comment (JeremyTheocharis) [inline golang/cmd/factoryinsight/database.go:1795]: Can you split it up into two files?

database.go to fetch all necessary data points and then dataprocessing.go to work with it? Then you can write a test for dataprocessing.go 

The test will be definitely necessary to ensure quality. Especially the splitting up of orders is in my experience very error prone and users do not tolerate any errors here ("if this is not correct, how can I trust the rest of the data?")

File: golang/cmd/factoryinsight/http.go (hunk 3 above)
Comment (Scarjit) [reply golang/cmd/factoryinsight/http.go:2082]: Inside database.go.
Every other http handle also does not specify this.

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]

File: golang/internal/regression.go (hunk 4 above)
Comment (Scarjit) [reply golang/internal/regression.go:1]: regression.go is used for the prediction inside GetAccumulatedProducts(...)

**Comment by JeremyTheocharis** [comment]
> > Did you push all your commits? This looks like it is not fully finished (only went through the code, did not test it)
> 
> This was finished, until @kruegeralex found some bugs

database.go was not shown in the Github Android App for unknown reasons. I saw it in the Desktop version and gave feedback

File: golang/cmd/factoryinsight/database.go (hunk 5 above)
Comment (Scarjit) [reply golang/cmd/factoryinsight/database.go:1841]: Fixed in 11be320b0bee8e43fe0298450b5c2e8fed8dcdaa

Comment (JeremyTheocharis) [reply golang/cmd/factoryinsight/database.go:1841]: The test is still open. Can you write some with also some edge cases (e.g., multiple orders, etc.)?

Comment (Scarjit) [reply golang/cmd/factoryinsight/database.go:1841]: Yes, im currently working on that

**Comment by Scarjit** [comment]
Superseded by #910
